extensions/amp-form/0.1/amp-form.js

```diff
@@ -177,8 +185,69 @@ export class AmpForm {
 
 
 /**
+ * Reports validity for the first invalid input - if any.
+ * @param {!HTMLFormElement} form
+ * @private
+ */
+function reportFormValidity(form) {
+  const inputs = toArray(form.querySelectorAll('input,select,textarea'));
```

Comment:
It's always `element.querySelectorAll('.class')` or such. The trick is that adding `:scope` to it is a lot faster. The problem: `:scope` is not available everywhere so we can't do it universally. An inconvenience for a generic solution is that you have to have `:scope` for each sub-selector. I.e. `input, select` becomes `:scope input, :scope select`.

Alternatively, you can use `element.getElementsByTagName()` API or such: obviously given three selectors this might not be better.